### PR TITLE
Normalization spec

### LINKED ISSUE #76: Interim normalized schema spec
## ⚠️ Problem
EA, MOSA, MCA, CSP, and classic Internal subscriptions all use different schemas. FOCUS is not available yet, so we need an interim schema.

## 🛠️ Solution
Standardize a cost data schema that can be used until FOCUS is finalized.

The following changes have been scoped out for now to minimize changes compared to the Cost Management connector:
- Rename AccountOwnerId to ClassicAccountOwnerId
- Rename AccountName to ClassicAccountName
- Rename BenefitId/Name to CommitmentId/Name
- Rename BillingCurrencyCode to Currency
- Rename ConsumedService to ResourceProvider = ConsumedService \|\| (parse from ResourceId)
- Rename CostInBillingCurrency to Cost
- Rename CustomerTenantId to CustomerId
- Rename CustomerName to CustomerName
- Rename MeterCategory to ServiceName
- Rename MeterSubCategory to ServiceTier
- Rename PartnerTenantId to PartnerId
- Rename ProductOrderId to OfferId
- Rename ProductOrderName to OfferName
- Rename ResellerMpnId to ResellerId
- Add ResourceGroupId = ResourceGroupName == null ? null : (parse from ResourceId)
- Add ResourceLocationId and ResourceLocationName for internal ID and English display names
- Add ResourceTypeName (display names from Azure portal)
- Add ServiceLevel = Allowed: `OnPremises`, `Infrastructure`, `Platform`, `Software`, `Other`
- Add Hybrid Benefit columns from Power BI reports: HybridBenefitStatus, HybridBenefitVCPUs, CPUHours, CPUs, ImageType, InstanceName, SKU, UsageType, VMProperties, vCPUs
- Add unique name columns from Power BI reports: BenefitNameUnique, ResourceGroupNameUnique, ResourceNameUnique, SubscriptionNameUnique
- Add utilization and savings columns from Power BI reports: BenefitSavings, BenefitType, BenefitUtilizationAmount, BenefitUtilizationPotential, DiscountSavings, NegotiatedSavings, OnDemandCost
- Consider merging ServiceInfo1 and ServiceInfo2 into AdditionalInfo
- Only show column PartnerEarnedCreditRate/Applied if available
- FOCUS: Rename ServiceFamily to ServiceCategory + add mapping to allowed values
- Consolidation of actual and amortized cost into a single dataset
- Consider a way to opt in to localized strings

## 🙋‍♀️ Ask for the community

We could use your help:
1. Please vote this issue up (👍) to prioritize it.
2. Leave comments to help us solidify the vision.

### PR DESCRIPTION
**Pull request by flanakin**
## 🛠️ Description
Spec to normalize EA, MCA, CSP, MOSA, and legacy internal cost data.

Fixes #76

### CODE REVIEW

File: docs/data-dictionary.md
Comment (flanakin) [inline docs/data-dictionary.md:1]: 💡 Tip: Hide whitespace when reviewing this file to make it easier to see the changes.
image

File: src/modules/pipelines/README.md [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+| BillingPeriodStartDate     | Date   |
+| BillingPeriodEndDate       | Date   |
+| ServicePeriodStartDate     | Date   |
+| ServicePeriodEndDate       | Date   |
+| Quantity                   | Number | Quantity \|\| UsageQuantity                                                                                 |
+| UnitPrice                  | Number | UnitPrice \|\| ResourceRate                                                                                 |
+| EffectivePrice             | Number | EffectivePrice \|\| UnitPrice                                                                               |
+| PayGPrice                  | Number | PayGPrice \|\| UnitPrice                                                                                    |
+| UnitOfMeasure              | String |
+| BillingCurrency            | String | BillingCurrency \|\| BillingCurrencyCode \|\| Currency                                                      |
```
Comment (tsilvers-ms) [inline src/modules/pipelines/README.md:142]: Can we normalize on {type}CurrencyCode instead? Also add to notes: ISO 4217 Currency Codes

File: src/modules/pipelines/README.md [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+| Cost                       | Number | Cost \|\| CostInBillingCurrency \|\| PreTaxCost                                                             |
```
Comment (tsilvers-ms) [inline src/modules/pipelines/README.md:143]: Should this match Billing? Normally when talking about money in different currencies, it's XXAmount for amount and XXCurrencyCode for the currency. So CostAmount and CostCurrencyCode or BillingAmount and BillingCurrencyCode. Big picture an easy way to match amount to currency via the names.

File: src/modules/pipelines/README.md [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+| Frequency                  | String |
+| CostAllocationRuleName     | String |
+| InvoiceId                  | String |
+| PreviousInvoiceId          | String |
+| BillingPeriodStartDate     | Date   |
+| BillingPeriodEndDate       | Date   |
+| ServicePeriodStartDate     | Date   |
+| ServicePeriodEndDate       | Date   |
+| Quantity                   | Number | Quantity \|\| UsageQuantity                                                                                 |
+| UnitPrice                  | Number | UnitPrice \|\| ResourceRate                                                                                 |
```
Comment (tsilvers-ms) [inline src/modules/pipelines/README.md:138]: Should our assumptions for currency be expanded here? Maybe explained in notes?

File: src/modules/pipelines/README.md [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+| PublisherType              | String | PublisherType \|\| "Azure"                                                                                  | Default to "Azure" for classic accounts.                |
+| PublisherId                | String |
+| PublisherName              | String | PublisherName \|\| "Microsoft"                                                                              | Default to "Microsoft" for classic accounts.            |
+| PlanName                   | String |
+| ServiceFamily              | String |
+| MeterCategory              | String | ServiceName \|\| MeterCategory                                                                              | Prefer ServiceName for EA.                              |
+| MeterSubCategory           | String | ServiceTier \|\| MeterSubCategory \|\| MeterSubcategory                                                     | Prefer ServiceTier for EA.                              |
+| ProductId                  | String |
```
Comment (tsilvers-ms) [inline src/modules/pipelines/README.md:114]: For values that are GUIDs should we add that to notes?

File: src/modules/pipelines/README.md [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+MG (Aug 2022):
+    DepartmentName, AccountName, AccountOwnerId, SubscriptionGuid, SubscriptionName, ResourceGroup, ResourceLocation, AvailabilityZone, UsageDateTime, ProductName, MeterCategory, MeterSubcategory, MeterId, MeterName, MeterRegion, UnitOfMeasure, UsageQuantity, ResourceRate, PreTaxCost, CostCenter, ConsumedService, ResourceType, InstanceId, Tags, OfferId, AdditionalInfo, ServiceInfo1, ServiceInfo2, Currency
+-->
+
+> ℹ️ _The table assumes all fields are Pascal-cased (first letter of each word capitalized)._
+
+| Column                     | Type   | Value                                                                                                       | Notes                                                   |
```
Comment (tsilvers-ms) [inline src/modules/pipelines/README.md:75]: Would love to also document maximum length for strings and format for Date, those are important for schema definitions.

**Pull request closed without merging**